pkg/util/objectwatcher/objectwatcher.go

```diff
@@ -172,10 +174,33 @@ func (o *objectWatcherImpl) Update(ctx context.Context, clusterName string, desi
 
 	desireObj, err = o.retainClusterFields(desireObj, clusterObj)
 	if err != nil {
-		klog.Errorf("Failed to retain fields for resource(kind=%s, %s/%s) in cluster %s: %v", clusterObj.GetKind(), clusterObj.GetNamespace(), clusterObj.GetName(), clusterName, err)
+		klog.Errorf("Failed to retain fields for desired resource(kind=%s, %s/%s) in cluster %s: %v", clusterObj.GetKind(), clusterObj.GetNamespace(), clusterObj.GetName(), clusterName, err)
 		return OperationResultNone, err
 	}
 
+	clusterObjCopy := clusterObj.DeepCopy()
+	err = prune.RemoveIrrelevantFields(clusterObjCopy, prune.RemoveJobTTLSeconds)
+	if err != nil {
+		klog.Errorf("Failed to remove irrelevant fields for cluster resource(kind=%s, %s/%s) in cluster %s: %v", clusterObj.GetKind(), clusterObj.GetNamespace(), clusterObj.GetName(), clusterName, err)
+		return OperationResultNone, err
+	}
+
+	// call retain interpret hook again.
+	clusterObjCopy, err = o.retainClusterFields(clusterObjCopy, clusterObj)
+	if err != nil {
+		klog.Errorf("Failed to retain fields for cluster resource(kind=%s, %s/%s) in cluster %s: %v", clusterObj.GetKind(), clusterObj.GetNamespace(), clusterObj.GetName(), clusterName, err)
+		return OperationResultNone, err
+	}
+
+	// If there's no actual content changes, skip the update and record the current version.
+	// if there's some strange mutating webhook in member cluster, theoretically this DeepEqual check will fail,
+	// so objectWatcher's VersionRecord is still needed.
```

Comment:
Can you elaborate on what kind of mutating logic?

Reply:
Let's take a most common example:

For native resources, objects will be mutated when entering karmada-apiserver, so when they are updated into member clusters by objectWatcher, they will perform the mutating logic again, which may be exactly the same with the one in karmada-apiserver, so deepequal could return true.

But for custom resources, when cluster administrators integrate karmada, they may only deploy mutating webhooks in member clusters, and desiredObject will not be mutated, while clusterObject is always mutated, so deepequal may always return false.

Reply:
In a very extreme case, adding deepequal may be harmful.

Suppose there is such a webhook in the member cluster, which will mutate the value of a field to 1 when the input is 0, and mutate it to 2 when the input is 1.

Then if the value of this field of this object in the member cluster is 1, the value of this field in desiredObject is also 1. Without deepequal, the update will succeed and change the value of the member cluster to 2. Then after adding deepequal, because deepequal will return true, the update is skipped, and this field of this object in the member cluster cannot be modified to 2.

Of course, if a mutatingwebhook is implemented in this way, I think it is very, very strange.